cli/src/movement_util.rs

```diff
@@ -0,0 +1,210 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::io::Write;
+use std::rc::Rc;
+
+use itertools::Itertools;
+use jj_lib::backend::CommitId;
+use jj_lib::commit::Commit;
+use jj_lib::repo::Repo;
+use jj_lib::revset::{RevsetExpression, RevsetFilterPredicate, RevsetIteratorExt};
+
+use crate::cli_util::{short_commit_hash, WorkspaceCommandHelper};
+use crate::command_error::{user_error, CommandError};
+use crate::ui::Ui;
+
+pub enum Direction {
+    Next,
+    Prev,
+}
+
+impl Direction {
+    fn cmd(&self) -> String {
+        match self {
+            Direction::Next => "next".to_string(),
+            Direction::Prev => "prev".to_string(),
+        }
+    }
+
+    fn target_not_found_message(&self, change_offset: u64) -> String {
+        match self {
+            Direction::Next => format!("No descendant found {} commit(s) forward", change_offset),
+            Direction::Prev => format!("No ancestor found {} commit(s) back", change_offset),
+        }
+    }
+
+    fn get_target_revset(
+        &self,
+        working_commit_id: &CommitId,
+        edit: bool,
+        has_conflict: bool,
+        change_offset: u64,
+    ) -> Result<Rc<RevsetExpression>, CommandError> {
+        let wc_revset = RevsetExpression::commit(working_commit_id.clone());
+        // If we're editing, start at the working-copy commit. Otherwise, start from
+        // its direct parent(s).
+        let start_revset = if edit {
+            wc_revset.clone()
+        } else {
+            wc_revset.parents()
+        };
+
+        let target_revset = match self {
+            Direction::Next => if has_conflict {
+                start_revset
+                    .children()
+                    .descendants()
+                    .filtered(RevsetFilterPredicate::HasConflict)
+                    .roots()
+            } else {
+                start_revset.descendants_at(change_offset)
+            }
+            .minus(&wc_revset),
+
+            Direction::Prev => {
+                if has_conflict {
+                    // If people desire to move to the root conflict, replace the `heads()` below
+                    // with `roots(). But let's wait for feedback.
+                    start_revset
+                        .parents()
+                        .ancestors()
+                        .filtered(RevsetFilterPredicate::HasConflict)
+                        .heads()
+                } else {
+                    start_revset.ancestors_at(change_offset)
+                }
+            }
+        };
+
+        Ok(target_revset)
+    }
+}
+
+fn get_target_commit(
+    ui: &mut Ui,
+    workspace_command: &WorkspaceCommandHelper,
+    direction: &Direction,
+    working_commit_id: &CommitId,
+    edit: bool,
+    has_conflict: bool,
+    change_offset: u64,
```

Comment:
nit: I would extract `edit, has_conflict, change_offset` to `struct MovementOptions` or something. 

(It could be `MovementArgs` flattened to `Next`/`PrevArgs`, but we would need to rephrase the help messages.)

Reply:
See #4297